wolfSSL/src/chapter06.md

```diff
@@ -185,3 +185,158 @@ These contexts may be pointers to any user-specified context, which will then in
+            ret = 0; /* or wolfCrypt error code */
+            break;
+
+        /* and so on for other algo types... */
+    }
+
+    return ret; // Return success or an appropriate error code
+}
+```
+
+Some of the simpler algo type unions, such as the RNG and seed unions, require no further processing and can be immediately acted on. However, more complicated operations like cipher or public key types have multiple union variants unto themselves that must be conditionally interpreted in the same manner. The variants and levels of hierarchy of each union are specific to each category of algorithm type, the full definitions of which can be found in the definition of the `wc_CryptoInfo` structure. As a general rule, each level that is a union will contain some sort of type indicator informing how the rest of the union should be dereferenced.
+
+Here is a simplified example for a complex algo type union with multiple levels of union decoding. The callback contains support for random number generation, as well as ECC key agreement, sign, and verify.
+
+```
+int myCryptoCallback(int devId, wc_CryptoInfo* info, void* ctx)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+
+    switch (info->algo_type) {
+        case WC_ALGO_TYPE_PK:
+            /* Handle public key operations, using algo type union: info->pk */
+            switch (info->pk.type) { /* pk type is of type wc_PkType */
+                case WC_PK_TYPE_ECDH:
+                    /* use info->pk.ecdh */
+                    ret = 0; /* or wolfCrypt error code */
+                    break;
+
+                case WC_PK_TYPE_ECDSA_SIGN:
+                    /* use info->pk.eccsign */
+                    ret = 0; /* or wolfCrypt error code */
+                    break;
+
+                case WC_PK_TYPE_ECDSA_VERIFY:
+                    /* use info->pk.eccverify */
+                    ret = 0; /* or wolfCrypt error code */
+                    break;
+            }
+            break;
+
+        case WC_ALGO_TYPE_RNG:
+            /* use info->rng */
+            ret = 0; /* or wolfCrypt error code */
+            break;
+    }
+
+    return ret;
+}
+```
+
+### Handling the request
+
+The data structure for each type of request generally contains a pointer and associated size for input and output data. Depending on the request it may include additional data including cryptographic keys, nonces, or further configuration. The crypto callback should operate on the input data and write relevant output data back to the appropriate variant of the algo type union. Writing data to a union variant that does not correspond to the algorithm type in question (e.g. using `info->cipher` when `info->algo_type == WC_ALGO_TYPE_RNG` is a memory error and can result in undefined behavior.
+
+
+### Examples
+
+Full examples of crypto callbacks can be found in the following locations
+
+    VaultIC Crypto Callbacks: https://github.com/wolfSSL/wolfssl-examples/blob/master/ccb_vaultic/ccb_vaultic.c
```

Comment:
Please use hyperlink markdown formatting. I also fixed the wolfTPM link

Full examples of crypto callbacks can be found in the following locations:

* [VaultIC Crypto Callbacks](https://github.com/wolfSSL/wolfssl-examples/blob/master/ccb_vaultic/ccb_vaultic.c)
* [STSAFE-A100 ECC Crypto Callbacks](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/port/st/stsafe.c)
* [TPM 2.0 wolfTPM Crypto Callbacks](https://github.com/wolfSSL/wolfTPM/blob/master/src/tpm2_cryptocb.c)
* [Generic wolfCrypt tests](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/test/test.c)

Reply:
done